### PR TITLE
[SPARK-36211][PYTHON] Correct typing of `udf` return value

### PR DESCRIPTION
**Pull request by luranhe**
The following code should type-check:

```python3
import uuid

import pyspark.sql.functions as F

my_udf = F.udf(lambda: str(uuid.uuid4())).asNondeterministic()
```

### What changes were proposed in this pull request?

The `udf` function should return a more specific type.

### Why are the changes needed?

Right now, `mypy` will throw spurious errors, such as for the code given above.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

This was not tested. Sorry, I am not very familiar with this repo -- are there any typing tests?

### CODE REVIEW

File: python/pyspark/sql/functions.pyi
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.pyi:363]: It returns a wrapped function. So the previous signature is correct.

Comment (HyukjinKwon) [reply python/pyspark/sql/functions.pyi:363]: `asNondeterministic` etc. is manually set to the specific wrapped function. Hm, we should fix the issue though

Comment (HyukjinKwon) [reply python/pyspark/sql/functions.pyi:363]: @zero323 do you have any thought on this?

Comment (zero323) [reply python/pyspark/sql/functions.pyi:363]: Yeah, this should be addressed, but I am not sure what is the best approach here. We'll definitely need a protocol  ‒ at first glance something like this should be sufficient:

```python
class UserDefinedFunctionWrapper(Protocol):
    def __call__(self, *args: Column) -> Column: ...
    def asNonDeterministic(self) -> UserDefinedFunctionWrapper: ...
```
and shouldn't break any existing annotations, but I haven't tested it thoroughly. 

@luranhe Thanks for your proposal. Would you mind creating a [JIRA ticket](https://issues.apache.org/jira/projects/SPARK/issues/) and updating the title accordingly (it should be of form `[SPARK-XXXXX][PYTHON] ...`). Thanks in advance.

Comment (HyukjinKwon) [reply python/pyspark/sql/functions.pyi]: Yeah, let's do that

Comment (zero323) [reply python/pyspark/sql/functions.pyi]: Also, `UserDefinedFunctionWrapper` is just a typing helper without corresponding class, so let's move it into `pyspark.sql._typing.pyi`.

**Comment by HyukjinKwon** [comment]
Jenkins, ok to test

File: python/pyspark/sql/functions.pyi
Comment (luranhe) [reply python/pyspark/sql/functions.pyi]: Just took a look, should we be using `UserDefinedFunctionLike` [here](https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/python/pyspark/sql/_typing.pyi#L56)?

Comment (zero323) [reply python/pyspark/sql/functions.pyi]: Yes, nice catch @luranhe!

**Comment by HyukjinKwon** [comment]
I'll leave it to @zero323.

File: python/pyspark/sql/functions.pyi
Comment (zero323) [inline python/pyspark/sql/functions.pyi:20]: Nitpick: We don't need `Protocol` here.

Comment (luranhe) [reply python/pyspark/sql/functions.pyi:20]: My bad, sorry

File: python/pyspark/sql/_typing.pyi
Comment (zero323) [inline python/pyspark/sql/_typing.pyi:62]: To be precise [wrapper uses](https://github.com/apache/spark/blob/2fe12a75206d4dbef6d7678b876c16876136cdd0/python/pyspark/sql/udf.py#L198) `*args` not `*cols`.  I used `_`, because what we advertise to the end user, is different from wrapped signature (we might be able to do better, once we drop support for Python < 3.10, but that's far away from now). If we want to change this, I'd recommend `*args` not `*cols`.

Other than that LGTM.

Comment (luranhe) [reply python/pyspark/sql/_typing.pyi:62]: Oh my bad, that's a good catch! However, if we change this to be `args` then a `UserDefinedFunction` wouldn't type properly as a `UserDefinedFunctionLike`, would it?

Comment (zero323) [reply python/pyspark/sql/_typing.pyi:62]: If I recall correctly only keyword argument names in `Protocols` are checked (and this can be explicitly disabled)\, but even if it wasn't the case, I don't think our intention was ever to cover `UserDefinedFunction` instances.

Comment (zero323) [reply python/pyspark/sql/_typing.pyi:62]: (Specifically, we needed a way to express a fact, that object that we get, is callable and takes variable number of `str` or `Column` arguments and returns a `Column`. This is something that cannot be done with simple `Callable`, but is quite useful for the end user)

**Review by zero323 — APPROVED** [review]

File: python/pyspark/sql/_typing.pyi
Comment (zero323) [reply python/pyspark/sql/_typing.pyi]: Thanks @luranhe ‒ I'll merge it tomorrow, unless there are any further comments.

**Review by HyukjinKwon — APPROVED** [review]
Looks good to me too. Let me leave it to @zero323.

**Comment by HyukjinKwon** [comment]
Test failure here https://github.com/luranhe/spark/runs/3162748315?check_suite_focus=true should be ignorable and unrelated.

**Comment by zero323** [comment]
Merged to master, branch-3.2 and branch-3.1.

**Pull request closed without merging**
